src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2794 @@
+
+    int i = 0;
+    switch (type.bytes()) {
+        case 1  : i = 0; break;
+        case 2  : i = 1; break;
+        case 4  : i = 2; break;
+        case 8  : i = 3; break;
+        default : halide_assert(user_context, false);  break;
+    }
+
+    DXGI_FORMAT format = DXGI_FORMAT_UNKNOWN;
+    format = FORMATS[(int)type.code][type.lanes-1][i];
+    return format;
+}
+
+#define INLINE inline __attribute__((always_inline))
+
+
+
+// The default implementation of halide_d3d12_get_symbol attempts to load
+// the D3D12 runtime shared library/DLL, and then get the symbol from it.
+static void *lib_d3d12  = NULL;
+static void *lib_D3DCompiler_47 = NULL;
+static void *lib_dxgi = NULL;
+
+struct LibrarySymbol {
+    template<typename T>
+    operator T () { return (T)symbol; }
+    void *symbol;
+
+    static LibrarySymbol get(void *user_context, void *lib, const char *name) {
+        void *s = d3d12_get_library_symbol(lib, name);
+        LibrarySymbol symbol = { s };
+        return symbol;
+    }
+};
+
+static PFN_D3D12_CREATE_DEVICE              D3D12CreateDevice           = NULL;
+static PFN_D3D12_GET_DEBUG_INTERFACE        D3D12GetDebugInterface      = NULL;
+static PFN_D3D12_SERIALIZE_ROOT_SIGNATURE   D3D12SerializeRootSignature = NULL;
+static PFN_D3DCOMPILE                       D3DCompile                  = NULL;
+static PFN_CREATEDXGIFACORY1                CreateDXGIFactory1          = NULL;
+
+#if defined(__cplusplus) && !defined(_MSC_VER)
+#if defined(__MINGW32__)
+#undef __uuidof
+#endif
+
+#define UUIDOF(T) REFIID __uuidof(const T&) { return  IID_ ## T; }
+
+UUIDOF(ID3D12Device)
+UUIDOF(ID3D12Debug)
+UUIDOF(ID3D12CommandQueue)
+UUIDOF(ID3D12CommandAllocator)
+UUIDOF(ID3D12CommandList)
+UUIDOF(ID3D12GraphicsCommandList)
+UUIDOF(ID3D12Resource)
+UUIDOF(ID3D12PipelineState)
+UUIDOF(ID3D12RootSignature)
+UUIDOF(ID3D12DescriptorHeap)
+UUIDOF(ID3D12Fence)
+UUIDOF(ID3D12QueryHeap)
+
+UUIDOF(IDXGIFactory1)
+UUIDOF(IDXGIAdapter1)
+UUIDOF(IDXGIOutput)
+#endif
+
+template<typename ID3D12Type>
+struct halide_d3d12_wrapper {
+    operator ID3D12Type *()    { return reinterpret_cast<ID3D12Type*>(this); }
+    ID3D12Type *operator -> () { return reinterpret_cast<ID3D12Type*>(this); }
+};
+
+template<typename ID3D12Type>
+struct halide_d3d12_deep_wrapper {
+    ID3D12Type *p;
+    operator ID3D12Type*    () { return p; }
+    ID3D12Type *operator -> () { return p; }
+};
+
+struct halide_d3d12compute_device : public halide_d3d12_wrapper<ID3D12Device> { };
+struct halide_d3d12compute_command_queue : public halide_d3d12_wrapper<ID3D12CommandQueue> { };
+
+namespace Halide { namespace Runtime { namespace Internal { namespace D3D12Compute {
+
+typedef halide_d3d12compute_device        d3d12_device;
+typedef halide_d3d12compute_command_queue d3d12_command_queue;
+
+struct d3d12_buffer {
+    ID3D12Resource *resource;
+    UINT sizeInBytes;
+    UINT offset;        // FirstElement
+    UINT elements;      // NumElements
+    UINT offsetInBytes;
```

Comment:
Is it necessary to store both `offset` and `offsetInBytes`?

Reply:
Unfortunately, yes; `offset` is in terms of buffer elements; due to strides, alignments, padding and/or cropping, `offsetInBytes` might not be easily inferred from `offset`. it would also require carrying formats/type descriptors around in most data structures.

Reply:
Please add this clarification as a comment.